nodejs/device/src/index.ts

```diff
@@ -0,0 +1,318 @@
+    if (!Array.isArray(obj.thief.serviceAcks)) {
+      this.rejectC2d(msg);
+      return console.warn('Issue with serviceAcks property. Rejecting.');
+    }
+    console.info(`Received serviceAckIds: ${obj.thief.serviceAcks}`);
+    obj.thief.serviceAcks.forEach(id => {
+      if (this.serviceAckWaitList.has(id)) {
+        this.serviceAckWaitList.delete(id);
+      }
+      else {
+        console.warn(`Received unknown serviceAckId: ${id}`);
+      }
+    });
+    this.completeC2d(msg);
+  }
+
+  private rejectC2d(msg) {
+    this.client.reject(msg).catch(e => {console.warn('Error when rejecting C2D message: ' + e.message)});
+  }
+
+  private completeC2d(msg) {
+    this.client.complete(msg).catch(e => {console.warn('Error when completing C2D message: ' + e.message)});
+  }
+
+  private registerMessageListener() {
+    this.client.on('message', msg => {
+      let obj;
+      try {
+        obj = JSON.parse(msg.getData());
+      }
+      catch {
+        this.rejectC2d(msg);
+        return console.warn('Failed to parse received C2D message. Rejecting.');
+      }
+
+      if (!obj.thief || obj.thief.runId !== runId || obj.thief.serviceInstanceId !== this.serviceInstanceId) {
+        this.rejectC2d(msg);
+        console.warn("C2D received, but it's not for us: %j. Rejecting.", obj);
+      }
+      else if (obj.thief.cmd === 'serviceAckResponse') {
+        this.handleServiceAckResponseMessage(obj, msg);
+      }
+      else if (obj.thief.cmd === 'testC2d') {
+        this.handleTestC2dMessage(obj, msg);
+      }
+      else {
+        this.rejectC2d(msg);
+        console.warn('Unknown command received: %j. Rejecting', obj);
+      }
+    });
+  }
+
+  private startC2dMessageSending(): Promise<void> {
+    const patch = {
+      thief: {
+        testControl: {
+          c2d: {
+            messageIntervalInSeconds: this.settings.receiveC2dIntervalInSeconds,
+            send: true
+          }
+        }
+      } 
+    }
+    console.info('Enabling C2D message testing: %j', patch);
+    return promisify(this.twin.properties.reported.update)(patch);
+  }
+
+  private startTelemetrySending() {
+    this.telemetrySendingInterval = setInterval(() => {
+      if (this.serviceAckWaitList.size > this.settings.sendMessageAllowedFailureCount) {
+        this.stopDeviceApp(`The number of service acks being waited on exceeds sendMessageAllowedFailureCount of ${this.settings.sendMessageAllowedFailureCount}.`);
+      }
+
+      const serviceAckId = uuidv4();
+      const message = new Message(JSON.stringify({
+        thief: {
+          cmd: 'serviceAckRequest',
+          serviceInstanceId: this.serviceInstanceId,
+          runId: runId,
+          serviceAckId: serviceAckId,
+          serviceAckType: 'telemetry'
+        }
+      }));
+      this.client.sendEvent(message)
+        .catch(() => {console.warn('Error sending message with serviceAckId: ' + serviceAckId)})
+        .finally(() => {this.serviceAckWaitList.add(serviceAckId)});
```

Comment:
Later on, we'll be counting these kinds of errors so we can report the count and fail the test if it gets too high

Reply:
Right now, the `serviceAckId` gets added to to the `serviceAckWaitList` set regardless of whether the sending is successful.

So it ends up being caught when the set is too big because (presumably) the service never sends sends back a C2D message with that GUID. 

Is that not what we want?